app/forms/supplementary/ProcedureCode.scala

```diff
@@ -17,22 +17,58 @@
 package forms.supplementary
 
 import forms.MetadataPropertiesConvertable
-import play.api.data.Forms.{seq, text}
+import play.api.data.Forms.{optional, text}
 import play.api.data.{Form, Forms}
 import play.api.libs.json.Json
 import utils.validators.FormFieldValidator._
 
-case class ProcedureCodes(
-  procedureCode: String, // max 4 alphanumeric characters
-  additionalProcedureCodes: Seq[String] // max 99 codes, each is max 3 alphanumeric characters
-) extends MetadataPropertiesConvertable {
+case class ProcedureCode(
+  procedureCode: Option[String], // max 4 alphanumeric characters
+  additionalProcedureCode: Option[String] // max 99 codes, each is max 3 alphanumeric characters
+)
+
+object ProcedureCode {
+  implicit val format = Json.format[ProcedureCode]
+
+  private val procedureCodeLength = 4
+  private val additionalProcedureCodeLength = 3
+
+  val mapping = Forms.mapping(
+    "procedureCode" -> optional(
+      text()
+        .verifying("supplementary.procedureCodes.procedureCode.error.empty", _.trim.nonEmpty)
+        .verifying(
+          "supplementary.procedureCodes.procedureCode.error.length",
+          isEmpty or hasSpecificLength(procedureCodeLength)
+        )
+        .verifying("supplementary.procedureCodes.procedureCode.error.specialCharacters", isAlphanumeric)
+    ),
+    "additionalProcedureCode" -> optional(
+      text()
+        .verifying(
+          "supplementary.procedureCodes.additionalProcedureCode.error.length",
+          isEmpty or hasSpecificLength(additionalProcedureCodeLength)
+        )
+        .verifying("supplementary.procedureCodes.additionalProcedureCode.error.specialCharacters", isAlphanumeric)
+    )
+  )(ProcedureCode.apply)(ProcedureCode.unapply)
+
+  val formId = "ProcedureCodes"
+
+  def form(): Form[ProcedureCode] = Form(mapping)
+}
+
+case class ProcedureCodesData(procedureCode: Option[String], additionalProcedureCodes: Seq[String])
```

Comment:
There is no need to have both case classes in the same file. I'd rather extract it to separate one.

Reply:
Yeah, but those classes are related to each other, so I think it should be in the same place